virtualizarr/readers/tiff.py

```
@@ -1,82 +1,197 @@
-            filepath,
-            loadable_variables=loadable_variables,
-            reader_options=reader_options,
-            drop_variables=drop_variables,
-            indexes=indexes,
+    @staticmethod
+    def open_virtual_dataset(
+        filepath: str,
+        group: str | None = None,
+        drop_variables: Iterable[str] | None = None,
+        loadable_variables: Iterable[str] | None = None,
+        decode_times: bool | None = None,
+        indexes: Mapping[str, xr.Index] | None = None,
+        virtual_backend_kwargs: Optional[dict] = None,
+        reader_options: Optional[dict] = None,
+    ) -> xr.Dataset:
+        if not group:
+            raise ValueError("Group is required for the TIFFVirtualBackend")
+        if not virtual_backend_kwargs:
+            raise ValueError(
+                "virtual_backend_kwargs is required for specifying the store and file_id"
+            )
+        store = virtual_backend_kwargs.get("store", None)
+        if not store:
+            raise ValueError(
+                f"Expected 'store' key in virtual_backend_kwargs, got {virtual_backend_kwargs}"
+            )
+        file_id = virtual_backend_kwargs.get("file_id", None)
+        if not file_id:
+            raise ValueError(
+                f"Expected 'file_id' key in virtual_backend_kwargs, got {virtual_backend_kwargs}"
+            )
+        manifest_store = TIFFVirtualBackend.open_virtual_zarr(
+            filepath=filepath,
             group=group,
-            decode_times=decode_times,
+            object_store=store,
+            file_id=file_id,
         )
 
-        return construct_virtual_dataset(
-            virtual_vars=virtual_vars,
-            loadable_vars=loadable_vars,
-            indexes=indexes,
-            coord_names=coord_names,
-            attrs=attrs,
+        # Convert to a Zarr store
+        return xr.open_dataset(
+            manifest_store, engine="zarr", consolidated=False, zarr_format=3
```

I'm confused - isn't this going to create a dataset with all loadable variables???

---

All the variables are "loadable", but not loaded unless you call something like .data on them.

---

> All the variables are "loadable", but not loaded unless you call something like .data on them.

That's what I mean - isn't this going to create a `Dataset` fully of lazy numpy arrays instead of `ManifestArrays`?

---

True, the `ManifestArray`s are be stored in `ManifestStore._manifest_group._manifest_dict` rather than in each `DataArray`. I'll need to give this some thought, and probably have a discussion about how to approach this. We could have a final step that replaces the lazy numpy arrays with ManifestArrays for non-loadable variables but that would require a lot more spaghetti in the `ManifestStore` internals.

---

Putting aside how majorly disruptive this would be, why must the data variables contain ManifestArrays, rather than indexing the ManifestArrays from the Zarr Store instance at write time? I'm thinking about whether we could use this approach, call `.load()` on loadable variables, then use isinstance(var, numpy.ndarray) and isinstance(var, xarray.core.indexing.MemoryCachedArray) after the concatenation, merging, etc. to determine whether the data or the VirtualChunkSpec should be stored in Icechunk.

If that's possible, we could provide all the functionality of both the Zarr and Xarray APIs. e.g., to_zarr now could be used to create a "native" zarr store from any archival file format, or virtualize.to_icechunk could create a virtual zarr store with some inlined data. if we are really concerned about accidental data loading we could break the Zarr API a bit to prevent loading any data not included in 'loadable_variables'.

---

> why must the data variables contain `ManifestArrays`, rather than indexing the `ManifestArrays` from the Zarr Store instance at write time?

Because you can't do arbitrary transformations on `ManifestArrays` like you can on numpy arrays. IIUC your approach (which I think can be summarized as "lazily fetching the `ManifestArray` corresponding to this numpy array) would very often lead to people unable to write virtual variables to the store at write time, because of some step they did earlier.

---

I'm going to experiment a bit. Because if we have a `loadable_arrays` attribute on the ManifestStore, we can raise an error if Xarray calls `.get()` on any of the arrays that aren't set as loadable, preventing any eager computations that would prevent serialization. So, it wouldn't be a full feature complete dataset but you'd get better error messages than those relating to ChunkManagers. Still may be a bad idea, but I think there's no harm in spending an afternoon trying it out.

---

My intuition is that this is just going to un-separate the concerns we want to separate, but I support messing around 😀